lambda/fanout_write_id_index_lambda.py

```diff
@@ -0,0 +1,158 @@
+            output = {'finished': True}
+            return output
+
+    # 'operation' is used to identify what failed when writing to the
+    # index deadletter queue.
+    fanout_subargs_common = {
+        'operation': 'write_id_index',
+        'id_index_table': event['config']['object_store_config']['id_index_table'],
+        's3_index_table': event['config']['object_store_config']['s3_index_table'],
+        'id_count_table': event['config']['object_store_config']['id_count_table'],
+        'cuboid_bucket': event['config']['object_store_config']['cuboid_bucket'],
+        'index_deadletter_queue': event['config']['object_store_config']['index_deadletter_queue'],
+        'id_index_new_chunk_threshold': event['config']['object_store_config']['id_index_new_chunk_threshold'],
+        'cuboid_object_key': event['cuboid_object_key'],
+        'version': event['version'],
+        # Optionally tell the Index.IdWriter to pause at startup.  This can be
+        # used as another way to allow Dynamo to scale up or to reduce the
+        # number of concurrent lambdas.
+        'wait_secs': 0
+    }
+
+    if len(event['ids']) > 0:
+        fanout_subargs = pack_ids_for_lambdas(event['ids'])
+    else:
+        fanout_subargs = event['sub_args']
+
+    max_execs = int(event['max_write_id_index_lambdas']) 
+
+    fanout_args = {
+        'operation': event['operation'],
+        'cuboid_object_key': event['cuboid_object_key'],
+        'version': event['version'],
+        'config': event['config'],
+        'fanout_params': event['fanout_params'],
+        # Only send ids once so we don't fanout indefinitely.
+        'ids': [],
+        'id_index_step_fcn': event['id_index_step_fcn'],
+        'sub_sfn': event['id_index_step_fcn'],
+        'sub_sfn_is_full_arn': True,
+        'common_sub_args': fanout_subargs_common,
+        'sub_args': fanout_subargs,
+        'max_concurrent': event['fanout_params']['max_concurrent'],
+        'rampup_delay': event['fanout_params']['rampup_delay'],
+        'rampup_backoff': event['fanout_params']['rampup_backoff'],
+        'status_delay': event['fanout_params']['status_delay'],
+        # Purposely emptying running list.  Limits on checking on running
+        # executions too small for the amount of Index.CuboidSupervisors that 
+        # will be simultaneously calling fanout_nonblocking().
+        'running': [],
+        'results': [],
+        'finished': False,
+        'max_write_id_index_lambdas': max_execs,
+        'wait_secs': 10
+    }
+
+    num_execs = get_lambda_execs_in_last_min()
+    if max_execs - num_execs < len(fanout_subargs) * 2:
+        # Don't fanout during this execution.  Wait for more lambda 
+        # availability.
+        fanout_args['wait_secs'] = 10 + randint(15, 25)
+        print('Too many lambdas, pausing fanout for {} secs.'.format(
+            fanout_args['wait_secs']))
+        return fanout_args
```

Comment:
managing lambda invocations manually seems fine, but has letting AWS manage this through reserved capacity been thought through?

Reply:
Yeah, but I think I still need to potentially slow things down here, so there's a greater chance of each step function succeeding w/o specifying a crazy number of retries. That being said, I have new code to push that does a "decorrelated" backoff. Step function retries don't allow you to specify a random element, so the current step functions all collide when trying to write the same id. Then they all wait about the same amount of time, and collide again! :)